user-ssh-builders/README.md

```diff
@@ -0,0 +1,142 @@
+# Give users SSH access to builders
+
+Author: [@FrostyX](https://github.com/frostyx)
+
+
+## Motivation
+
+Users sometimes need SSH access to a builder instance to debug Copr-only
+build failures. We want them to be able to get the access without
+admin intervention.
+
+See also https://github.com/fedora-copr/copr/issues/2364
+
+
+## Limits
+
+Obviously, Copr resources aren't unlimited so we need to allocate them
+wisely.
+
+- Only 1 running instance per user
+    - Having 1 instance per project would be too much
+    - Having more than 1 instance per user isn't needed, users can
+      team up for debugging issues that need more builders at once
+    - No need for group instances, users can do whatever they want,
+      including adding access to their team members
+
+- Builders need to have a limited lifespan, e.g. 48 hours
+    - When submitting a build in Copr, we allow to adjust the timeout
+      to 30 hours at most. That means the lifespan needs to be at
+      least 30 hours.
+    - I propose 48 hours because we want to give users enough time for
+      debugging after the build finishes
+    - Alternatively, we can do a 24-hours lifespan and add an "extend"
+      button to Copr WebUI or to add a `copr-builder-extend` command
+      to the instance.
+
+- Users can do whatever they want with the builder instance ... but
+  not really. Users cannot violate any rules defined in
+  [What I can build in Copr?][what-can-i-build] and cannot use the
+  instance for torrents, exploits, bitcoin mining, ... I don't want to
+  give you ideas ... and neither for a personal VPS. We should print
+  some fancy legal text and make users agree with it by clicking the
+  button to obtain the login instructions.
+
+
+## Workflow
+
+### Option 1 - Button
+
+For failed builds, there could be a button on the build detail
+page. Clicking it would show the information about this feature,
+rules, instance type selection, etc., and a button to actually spawn
+the instance. The instance would be fresh and it would be up to the
+user to reproduce the failure manually. Once the builder spawns, the
+page will show instructions on how to connect to it.
+
+### Option 2 - Keep alive after failure
+
+There could be a checkbox when submitting a build (or alternatively
+in project/package/chroot settings) to keep the builder alive after
+the build fails. This option seems easier from the user perspective
+but:
+
+- It will probably be confusing when using the checkbox while some
+  instance is already running. We don't want to kill it and spawn a
+  new one. We probably need to print it was not possible to spawn a
+  new instance into the builder-live/backend log
+- Where to print the instructions on how to connect to the builder?
```

Comment:
The `backend.log` (backend is responsible for (not)closing the resalloc ticket)?